[inductor] make non-trivial tile ranges unbacked symint aware

ColinPeppler opened the pull request:
# New PR skipping this if unbackeds present: https://github.com/pytorch/pytorch/pull/150225

### Stacktrace
```
# If no fallback is provided then we'd see this.
  File "/data/users/colinpeppler/pytorch/torch/_inductor/codegen/simd.py", line 1746, in tile_ranges
    if V.graph.sizevars.atomically_apply_size_hint(
  File "/data/users/colinpeppler/pytorch/torch/_inductor/sizevars.py", line 653, in atomically_apply_size_hint
    size_dict = {
  File "/data/users/colinpeppler/pytorch/torch/_inductor/sizevars.py", line 654, in <dictcomp>
    symbol: V.graph.sizevars.size_hint(symbol, fallback=fallback)
  File "/data/users/colinpeppler/pytorch/torch/_inductor/sizevars.py", line 556, in size_hint
    return int(out)
  File "/home/colinpeppler/local/miniconda3/envs/pytorch/lib/python3.10/site-packages/sympy/core/expr.py", line 307, in __int__
    raise TypeError("Cannot convert symbols to int")
torch._inductor.exc.InductorError: TypeError: Cannot convert symbols to int
```

### Context
When unbacked symint is in the ranges, Inductor will choke up because either: 
1. no size hint exists or 
~~2. it can't statically evaluate an expression.~~

### Approach
- For (1), use unbacked symint fallback; this isn't optimal and we could do better on choosing a fallback
~~- For (2), use `statically_known*` mechanisms~~

xref: https://fb.workplace.com/groups/1028545332188949/permalink/1178385333871614/

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #149994



cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov

ColinPeppler committed:
[inductor] make non-trivial unbacked symint aware

[ghstack-poisoned]

ColinPeppler commented:
@pytorchbot rebase

ColinPeppler committed:
Update

[ghstack-poisoned]

jingsh reviewed (approved):
lgtm

ColinPeppler commented:
`test_dynamic_shapes_2d_blocking` is failing, gonna take a look

blaine-rister reviewed (approved):
LGTM! Nice fix.

eellison reviewed (changes requested):
please get a review of core inductor folks before landing changes to sensitive tiling code like this.

eellison reviewed (commented):
I would like to understand the problem more. Generally we use size hints throughout scoring (e.g. in fusions). Often sympy will not be able to prove equivalence so hints can be more robust. 

If there error is just about converting sympy expr -> int I think we're missing a hint call somewhere

ColinPeppler commented:
My bad! was just about to request you elias

blaine-rister commented on torch/_inductor/codegen/simd.py:
```diff
@@ -1736,11 +1737,8 @@ def collapse_ranges(ranges: Sequence[sympy.Expr]) -> sympy.Expr:
-                        score - sympy_product(itertools.chain(ranges, reduction_ranges))
-                    )
-                    >= 0
+                if V.graph.sizevars.statically_known_geq(
+                    score, sympy_product(itertools.chain(ranges, reduction_ranges))
                 ):
```
Whoops, good catch. Admittedly I'm not too familiar with the limitations of statically_known vs. size_hints.

ColinPeppler replied:
@eellison Ok will change back to size-hint, but I want to understand a bit more on when to use `statically_known*` versus `size_hint` because I kept thinking to myself that `statically_known*` made more sense here.

From what I understand, `size_hint` is when you believe the expr in-hand has size-like symbols are you want use some literal value. So, I can see it fitting here.

However, my intuition was that it is better to skip an optimization due to size uncertainty from unbacked symints, instead of making an optimization decision that may not turnout to be optimal (I feel that size_hint with fallback does this).

ColinPeppler committed:
Update on "[inductor] make non-trivial tile ranges unbacked symint aware"

### Stacktrace
```
# If no fallback is provided then we'd see this.
  File "/data/users/colinpeppler/pytorch/torch/_inductor/codegen/simd.py", line 1746, in tile_ranges
    if V.graph.sizevars.atomically_apply_size_hint(
  File "/data/users/colinpeppler/pytorch/torch/_inductor/sizevars.py", line 653, in atomically_apply_size_hint
    size_dict = {
  File "/data/users/colinpeppler/pytorch/torch/_inductor/sizevars.py", line 654, in <dictcomp>
    symbol: V.graph.sizevars.size_hint(symbol, fallback=fallback)
  File "/data/users/colinpeppler/pytorch/torch/_inductor/sizevars.py", line 556, in size_hint
    return int(out)
  File "/home/colinpeppler/local/miniconda3/envs/pytorch/lib/python3.10/site-packages/sympy/core/expr.py", line 307, in __int__
    raise TypeError("Cannot convert symbols to int")
torch._inductor.exc.InductorError: TypeError: Cannot convert symbols to int
```

### Context
When unbacked symint is in the ranges, Inductor will choke up because either: 
1. no size hint exists or 
2. it can't statically evaluate an expression.

### Approach
- For (1), use unbacked symint fallback; this isn't optimal and we could do better on choosing a fallback
- For (2), use `statically_known*` mechanisms

xref: https://fb.workplace.com/groups/1028545332188949/permalink/1178385333871614/




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

ColinPeppler commented on torch/_inductor/codegen/simd.py:
```diff
@@ -2110,8 +2107,9 @@ class CandidateTiling:
         """Somewhat arbitrary heuristic used to boost scores for some sizes"""
-        s = V.graph.sizevars.size_hint(s)
-        return s >= 32 and (s % 32 == 0)
+        return V.graph.sizevars.statically_known_geq(
```
Just sharing what I was thinking when using `statically_known` here, it's because `config.unbacked_symint_fallback = 8192` which is always divisible by 32 and greater than 32. 

In reality, the unbacked symint could be less than 32 or odd.

ColinPeppler commented:
@pytorchbot rebase

eellison commented:
all good - it's possible we do need static somewhere.. could you link to the exact stack trace / error you're seeing ?

ColinPeppler committed:
Update

[ghstack-poisoned]

ColinPeppler commented:
@eellison here is more context into the issue

An internal model is unable to do `aot_compile` because it'll run into candidate_tiling and choke on the size_hint. Here's a few example stacktraces.
```
  File "/data/users/colinpeppler/pytorch/torch/_inductor/codegen/simd.py", line 1320, in codegen_node
    return self.codegen_node_schedule(
  File "/data/users/colinpeppler/pytorch/torch/_inductor/codegen/simd.py", line 1354, in codegen_node_schedule
    tiling = self.select_tiling(
  File "/data/users/colinpeppler/pytorch/torch/_inductor/codegen/simd.py", line 1992, in select_tiling
    for candidate_tiling in cls.candidate_tilings(node, numel, reduction_numel):
  File "/data/users/colinpeppler/pytorch/torch/_inductor/codegen/simd.py", line 1772, in candidate_tilings
    partial_tilings = tile_ranges(
  File "/data/users/colinpeppler/pytorch/torch/_inductor/codegen/simd.py", line 1743, in tile_ranges
    if CandidateTiling.is_good_size(tiled_groups[1]):
  File "/data/users/colinpeppler/pytorch/torch/_inductor/codegen/simd.py", line 2118, in is_good_size
    s = V.graph.sizevars.size_hint(s)
   ...
torch._inductor.exc.InductorError: TypeError: Cannot convert symbols to int


  File "/data/users/colinpeppler/pytorch/torch/_inductor/codegen/simd.py", line 1320, in codegen_node
    return self.codegen_node_schedule(
  File "/data/users/colinpeppler/pytorch/torch/_inductor/codegen/simd.py", line 1354, in codegen_node_schedule
    tiling = self.select_tiling(
  File "/data/users/colinpeppler/pytorch/torch/_inductor/codegen/simd.py", line 1992, in select_tiling
    for candidate_tiling in cls.candidate_tilings(node, numel, reduction_numel):
  File "/data/users/colinpeppler/pytorch/torch/_inductor/codegen/simd.py", line 1772, in candidate_tilings
    partial_tilings = tile_ranges(
  File "/data/users/colinpeppler/pytorch/torch/_inductor/codegen/simd.py", line 1743, in tile_ranges
    if CandidateTiling.is_good_size(tiled_groups[1]):
  File "/data/users/colinpeppler/pytorch/torch/_inductor/codegen/simd.py", line 2118, in is_good_size
    s = V.graph.sizevars.size_hint(s)
  File "/data/users/colinpeppler/pytorch/torch/_inductor/sizevars.py", line 556, in size_hint
    return int(out)
  File "/home/colinpeppler/local/miniconda3/envs/pytorch/lib/python3.10/site-packages/sympy/core/expr.py", line 307, in __int__
    raise TypeError("Cannot convert symbols to int")
torch._inductor.exc.InductorError: TypeError: Cannot convert symbols to int


  File "/data/users/colinpeppler/pytorch/torch/_inductor/codegen/simd.py", line 2015, in convert_tiling_to_3d
    if V.graph.sizevars.size_hint(a1 - b1) == 0:
  File "/data/users/colinpeppler/pytorch/torch/_inductor/sizevars.py", line 556, in size_hint
    return int(out)
  File "/home/colinpeppler/local/miniconda3/envs/pytorch/lib/python3.10/site-packages/sympy/core/expr.py", line 307, in __int__
    raise TypeError("Cannot convert symbols to int")
torch._inductor.exc.InductorError: TypeError: Cannot convert symbols to int
```

In the case of tiled reductions, even if we try to disable by`config.triton.tile_reductions=False` or `config.triton.max_tiles <= -1`, it still ends up choking on the size_hint.
https://github.com/pytorch/pytorch/blob/6aca002d82e5131cbf48496a04e7b0213ace1c03/torch/_inductor/codegen/simd.py#L1963-L1971

So, that sprung me to make `candidate_tilings` unbacked symint aware. An alternative is to skip `candidate_tilings` if there's any unbacked symints present.

ColinPeppler committed:
Update on "[inductor] make non-trivial tile ranges unbacked symint aware"

### Stacktrace
```
# If no fallback is provided then we'd see this.
  File "/data/users/colinpeppler/pytorch/torch/_inductor/codegen/simd.py", line 1746, in tile_ranges
    if V.graph.sizevars.atomically_apply_size_hint(
  File "/data/users/colinpeppler/pytorch/torch/_inductor/sizevars.py", line 653, in atomically_apply_size_hint
    size_dict = {
  File "/data/users/colinpeppler/pytorch/torch/_inductor/sizevars.py", line 654, in <dictcomp>
    symbol: V.graph.sizevars.size_hint(symbol, fallback=fallback)
  File "/data/users/colinpeppler/pytorch/torch/_inductor/sizevars.py", line 556, in size_hint
    return int(out)
  File "/home/colinpeppler/local/miniconda3/envs/pytorch/lib/python3.10/site-packages/sympy/core/expr.py", line 307, in __int__
    raise TypeError("Cannot convert symbols to int")
torch._inductor.exc.InductorError: TypeError: Cannot convert symbols to int
```

### Context
When unbacked symint is in the ranges, Inductor will choke up because either: 
1. no size hint exists or 
~~2. it can't statically evaluate an expression.~~

### Approach
- For (1), use unbacked symint fallback; this isn't optimal and we could do better on choosing a fallback
~~- For (2), use `statically_known*` mechanisms~~

xref: https://fb.workplace.com/groups/1028545332188949/permalink/1178385333871614/




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

ColinPeppler committed:
Update on "[inductor] make non-trivial tile ranges unbacked symint aware"

### Stacktrace
```
# If no fallback is provided then we'd see this.
  File "/data/users/colinpeppler/pytorch/torch/_inductor/codegen/simd.py", line 1746, in tile_ranges
    if V.graph.sizevars.atomically_apply_size_hint(
  File "/data/users/colinpeppler/pytorch/torch/_inductor/sizevars.py", line 653, in atomically_apply_size_hint
    size_dict = {
  File "/data/users/colinpeppler/pytorch/torch/_inductor/sizevars.py", line 654, in <dictcomp>
    symbol: V.graph.sizevars.size_hint(symbol, fallback=fallback)
  File "/data/users/colinpeppler/pytorch/torch/_inductor/sizevars.py", line 556, in size_hint
    return int(out)
  File "/home/colinpeppler/local/miniconda3/envs/pytorch/lib/python3.10/site-packages/sympy/core/expr.py", line 307, in __int__
    raise TypeError("Cannot convert symbols to int")
torch._inductor.exc.InductorError: TypeError: Cannot convert symbols to int
```

### Context
When unbacked symint is in the ranges, Inductor will choke up because either: 
1. no size hint exists or 
~~2. it can't statically evaluate an expression.~~

### Approach
- For (1), use unbacked symint fallback; this isn't optimal and we could do better on choosing a fallback
~~- For (2), use `statically_known*` mechanisms~~

xref: https://fb.workplace.com/groups/1028545332188949/permalink/1178385333871614/




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

blaine-rister commented:
> @eellison here is more context into the issue
> 
> An internal model is unable to do `aot_compile` because it'll run into candidate_tiling and choke on the size_hint. Here's a few example stacktraces.
> 
> ```
>   File "/data/users/colinpeppler/pytorch/torch/_inductor/codegen/simd.py", line 1320, in codegen_node
>     return self.codegen_node_schedule(
>   File "/data/users/colinpeppler/pytorch/torch/_inductor/codegen/simd.py", line 1354, in codegen_node_schedule
>     tiling = self.select_tiling(
>   File "/data/users/colinpeppler/pytorch/torch/_inductor/codegen/simd.py", line 1992, in select_tiling
>     for candidate_tiling in cls.candidate_tilings(node, numel, reduction_numel):
>   File "/data/users/colinpeppler/pytorch/torch/_inductor/codegen/simd.py", line 1772, in candidate_tilings
>     partial_tilings = tile_ranges(
>   File "/data/users/colinpeppler/pytorch/torch/_inductor/codegen/simd.py", line 1743, in tile_ranges
>     if CandidateTiling.is_good_size(tiled_groups[1]):
>   File "/data/users/colinpeppler/pytorch/torch/_inductor/codegen/simd.py", line 2118, in is_good_size
>     s = V.graph.sizevars.size_hint(s)
>    ...
> torch._inductor.exc.InductorError: TypeError: Cannot convert symbols to int
> 
> 
>   File "/data/users/colinpeppler/pytorch/torch/_inductor/codegen/simd.py", line 1320, in codegen_node
>     return self.codegen_node_schedule(
>   File "/data/users/colinpeppler/pytorch/torch/_inductor/codegen/simd.py", line 1354, in codegen_node_schedule
>     tiling = self.select_tiling(
>   File "/data/users/colinpeppler/pytorch/torch/_inductor/codegen/simd.py", line 1992, in select_tiling
>     for candidate_tiling in cls.candidate_tilings(node, numel, reduction_numel):
>   File "/data/users/colinpeppler/pytorch/torch/_inductor/codegen/simd.py", line 1772, in candidate_tilings
>     partial_tilings = tile_ranges(
>   File "/data/users/colinpeppler/pytorch/torch/_inductor/codegen/simd.py", line 1743, in tile_ranges
>     if CandidateTiling.is_good_size(tiled_groups[1]):
>   File "/data/users/colinpeppler/pytorch/torch/_inductor/codegen/simd.py", line 2118, in is_good_size
>     s = V.graph.sizevars.size_hint(s)
>   File "/data/users/colinpeppler/pytorch/torch/_inductor/sizevars.py", line 556, in size_hint
>     return int(out)
>   File "/home/colinpeppler/local/miniconda3/envs/pytorch/lib/python3.10/site-packages/sympy/core/expr.py", line 307, in __int__
>     raise TypeError("Cannot convert symbols to int")
> torch._inductor.exc.InductorError: TypeError: Cannot convert symbols to int
> 
> 
>   File "/data/users/colinpeppler/pytorch/torch/_inductor/codegen/simd.py", line 2015, in convert_tiling_to_3d
>     if V.graph.sizevars.size_hint(a1 - b1) == 0:
>   File "/data/users/colinpeppler/pytorch/torch/_inductor/sizevars.py", line 556, in size_hint
>     return int(out)
>   File "/home/colinpeppler/local/miniconda3/envs/pytorch/lib/python3.10/site-packages/sympy/core/expr.py", line 307, in __int__
>     raise TypeError("Cannot convert symbols to int")
> torch._inductor.exc.InductorError: TypeError: Cannot convert symbols to int
> ```
> 
> In the case of tiled reductions, even if we try to disable by setting`config.triton.tile_reductions=False` or `config.triton.max_tiles <= -1`, it still ends up choking on the size_hint.
> 
> https://github.com/pytorch/pytorch/blob/6aca002d82e5131cbf48496a04e7b0213ace1c03/torch/_inductor/codegen/simd.py#L1963-L1971
> 
> So, that sprung me to make `candidate_tilings` unbacked symint aware. An alternative is to skip `candidate_tilings` if there's any unbacked symints present.

@ColinPeppler I'm not sure if this helps, but if we to encourage tiling with `config.triton.prefer_nd_tiling = True`, that often ends up bypassing the tiling heuristic from `CandidateTilings`. It still runs `CandidateTilings`, but if it finds its own ND candidates, those get ranked above the ones returned by `CandidateTilings`. So for your use case, it may not be essential to make `CandidateTilings` compatible with unbacked symints. It might work just as well to skip it.

@eellison  alternatively, is there a way to make Inductor generate a backed symint here? It seems like we should be able to bound the output shape of `permute` based on its input shape and `dims`. In the worst case, if we don't know `dims`, could we take the max across all dimensions? Sorry if this is a naive question.

ColinPeppler commented:
@blaine-rister I think it does help, it makes sense to skip optimization if unbackeds are present for now. Will take you up on the offer :)

I believe an ideal world is where non-trivial tiling can work with unbacked symints, but I feel that it isn't clear how to do that responsibly, and so, I should come up with some real test cases.

bobrenjc93 commented on torch/_inductor/lowering.py:
```diff
@@ -1170,9 +1170,12 @@ def inner_fn(index):
                     index[i] = ModularIndexing(index[i], 1, old_size[i])
         return x_loader(index)
 
+    from torch.fx.experimental.symbolic_shapes import free_unbacked_symbols
```
Is there a reason you imported this here? Is it for cycle breaking? These local imports can cause unwanted compile times: https://fb.workplace.com/groups/pytorch.dev/permalink/1685878491990624

ColinPeppler replied:
Sorry! this was out of a bad habit. There shouldn't be any cycle here, let me move it up to module-level.

eellison commented on torch/_inductor/codegen/simd.py:
```diff
@@ -2116,8 +2134,8 @@ class CandidateTiling:
         """Somewhat arbitrary heuristic used to boost scores for some sizes"""
-        s = V.graph.sizevars.size_hint(s)
-        return s >= 32 and (s % 32 == 0)
+        s = V.graph.sizevars.size_hint(s, fallback=config.unbacked_symint_fallback)
+        return s >= 32 and s % 32 == 0
```
Seems more conservative to fallback to the size not being a good size ?

ColinPeppler replied:
yeah that makes more sense, I'll make sure to do that when I revisit it

eellison commented on torch/_inductor/codegen/simd.py:
```diff
@@ -2008,20 +2010,36 @@ def select_tiling(
+                    )
+                    > 0
+                )
+                if not V.graph.sizevars.size_hint(
+                    a1 % b1 == 0, fallback=config.unbacked_symint_fallback
+                ):
```
This seems like something that should be statically_known..

eellison reviewed (commented):
i think when we want to do hints/statically known is a little subtle, lets just do the skip 3d with unbacked fo rnow
